lib/spack/spack/util/executable.py

```diff
@@ -269,39 +271,39 @@ def __str__(self):
         return " ".join(self.exe)
 
 
-@system_path_filter
 def which_string(*args, **kwargs):
     """Like ``which()``, but return a string instead of an ``Executable``."""
     path = kwargs.get("path", os.environ.get("PATH", ""))
     required = kwargs.get("required", False)
 
-    if isinstance(path, str):
-        path = path.split(os.pathsep)
+    if isinstance(path, list):
+        paths = [Path(str(x)) for x in path]
 
-    for name in args:
-        win_candidates = []
-        if sys.platform == "win32" and (not name.endswith(".exe") and not name.endswith(".bat")):
-            win_candidates = [name + ext for ext in [".exe", ".bat"]]
-        candidate_names = [name] if not win_candidates else win_candidates
+    if isinstance(path, str):
+        paths = [Path(x) for x in path.split(os.pathsep)]
 
+    for search_item in args:
+        is_cwd = True if search_item.startswith(".") else False
+        search_item = Path(search_item)
+        candidate_items = [Path(search_item)]
         if sys.platform == "win32":
-            new_path = path[:]
-            for p in path:
-                if os.path.basename(p) == "bin":
-                    new_path.append(os.path.dirname(p))
-            path = new_path
-
-        for candidate_name in candidate_names:
-            if os.path.sep in candidate_name:
-                exe = os.path.abspath(candidate_name)
-                if os.path.isfile(exe) and os.access(exe, os.X_OK):
-                    return exe
-            else:
-                for directory in path:
-                    directory = path_to_os_path(directory).pop()
-                    exe = os.path.join(directory, candidate_name)
-                    if os.path.isfile(exe) and os.access(exe, os.X_OK):
-                        return exe
+            if not search_item.suffix:
+                candidate_items = [
+                    search_item.parent / (search_item.name + ext) for ext in [".exe", ".bat"]
+                ]
```

Comment:
Like before, this is taking a `.../some/path/bin/foo` and also looking for `.../some/path/foo`. It adds more logic though, to specifically look for a `.exe`. The odd thing though is that it doesn't look for just `foo` in the parent, only `foo.exe` or `foo.bat`, so it's willing to find `[bin/foo, foo.exe, foo.bat]` which seems like an odd combination of things to accept; I would assume something like `[bin/foo, bin/foo.exe, bin/foo.bat, foo.bat, foo.exe, foo]`.

Reply:
The old logic did look for a `.exe` and a `.bat` [here](https://github.com/spack/spack/pull/37762/files/2062d63066e17ee9eb97f4935c31910905f93d84#diff-950806535b61829c748ff7d908c225e5b15a72cf0032354343ec4d58325ac6e8L284). I think that the current implementation searches for the full set you described at the end. I refactored it so it's a bit clearer to me now. Let me know if you still think the issue is still there.